### ISSUE TITLE
How do I avoid writing feature files and inject/load features dynamically?

### ISSUE THREAD

**Issue by ArielZamparini-REISys** [Behat/Behat#873]
Having trouble wrapping my head around this concept. I have features being created by a front end CMS application, I can modify those features through a web browser and do all my work through this interface. When I want the backend server to run my full test suite of features I have to create files for behat to see, then run behat process and have my test output. However is there a way of skipping the features files entirely? Such as telling behat to run a test that is injected into it dynamically... say from a database?

In the same line of questioning, is there a way to in the same regard load behat so that it can be run without console commands and basically use it as a PHP Library?

**Comment by jakubzapletal** [comment]
Would you consider to not keep features in DB? One of the points of feature files is a living documentation following the code. Keeping the features in DB means that it is out of version control of the code, and, if I understand properly, it is managed by another system, in that case by some front-end CMS.

**Comment by ArielZamparini-REISys** [comment]
Yes you are correct, I will not fight you there. My main use case for the front-end is to be able to have a team that does not have code base access to create, write and manipulate these test cases (a mix of business analysts, quality assurance, frontend and backend developers). We can have the ultimate goal being that we commit the correct test cases to code but having a whole team try and install all the required binaries and libraries correctly seemed like a more daunting task vs having them just use a website. The site in question has a built in syntax helper, allows the user to fill in the user stories and tracks pass/fail rates. 

Behat has been monumental in accomplishing this with how great everything works but it was becoming cumbersome to herd files all the time, or to invoke behat using execs().

**Comment by jakubzapletal** [comment]
Could you please clarify a use case when 2 developers are working on
different branches and features? Where do they get the data with features?
Do they have same features or different ones? How would you run CI?
Generally I wonder how would the agile development process look like.

**Comment by ArielZamparini-REISys** [comment]
We have a simplified Agile process (stand-ups, scrums, sprints, etc) and user stories are directly available in ticket tracking. Currently our issue is vastly different development systems... some users have virtual machines, some have mamp, some have wamp. Rather than put the development team on creating installation scripts and debug vastly different architectures we decided to build an app using the fabulous features of behat. You can extend it greatly with selenium automation, unit testing frameworks and all kinds of tools. Our eventual goal is to have jenkins (along with the jenkins pipeline plugins) run our CI process for us do tests before builds. At this point we have not reached these goals. We are in the process of getting people used to gherkin formatting and behat syntax, this is mainly part of adoption. 

The webapp we built helps greatly with this in creating the tests themselves. They can be easily exported and stamped out to version control to eventually become part of the process. The webapp can then simply receive jenkins callbacks to run the full suite of tools.

**Comment by jakubzapletal** [comment]
By the use case I mean if you have a repository with the main app, where do
you have your data related to the new branch? Do you have one central DB
for all developers and whole development process or do you somehow verse
your data with each branch?

**Comment by ArielZamparini-REISys** [comment]
We deploy one db and webserver per application along with associated services and configs. With the webapp I am speaking of it can quickly select that selenium connect to different urls to be able to test any phase of development whether that be dev, uat, qa, staging etc. With the data gathering abilities of the db we can see where in the process that a test failed much like most automated testing tools out there.

**Comment by scaytrase** [comment]
On the one hand - handling features from the other source would be nice. Something like

```sh
echo_some_dot_feature_content() | behat -
```

On the other hand behat feature is not the only scenario, but also the `Context`. In this case you should either provide context classes alongside your feature context (but nobody wants to known which classes, as your features are stored elsewhere) or use some kind of default context. Second means you will include almost ALL context classes to defaults and won't be able to use different context classes for the same steps (i.e. for different roles testing).

I've understood the main idea of your system and I'd better look into changing the storage, i.e store features directly in dedicated git repo (skipping the DB), like online github file editing. Just commit file on the save button click. This also immediatly achievs all benefits from VCS (authorship, history, branching)

**Comment by ArielZamparini-REISys** [comment]
Currently the file stamping mechanism I have saves the features from the db write out to direct files that are stored in a directory that behat can run on. Each feature knows to which client they belong so my folder structure ends up being

+-- /var/tests/behat
|   +-- behat binaries
|   +-- features
|   |   +-- client_1
|   |   |   +-- context_file_for_client_1
|   |   |   +-- feature_tests_for_client_1
etc...

Again no problems with behat or the eventual goal of making features part of projects, I understand while you would want to keep that mantra for the project. Just looking for flexibility to running behat within a php environement without console. Otherwise people look at me funny for using exec for running behat T_T

**Comment by djdevin** [comment]
We do the same thing. QA does not have access to Git for various reasons and that probably is not going to change. We use Testrail for management of manual tests (a lot) so that's where the automated tests lie as well. This is an "enterprise" system so every test case has to have an ID and some other businessy data associated with it - requirement documents, credentials, etc. Testrail has revisioning and branching, so we really have no reason to move everything into .feature files. Sometimes QA can write failing tests intentionally if it's a release blocker and we know what the expectation is - like a message appearing or some field containing a value. Then when the developer creates their feature branch to fix it, it gets tested against the current automated tests and if it passes, it gets merged in.

Piping really should be supported since Gherkin seems to be geared towards non-developers. Our QA writes the majority of Gherkin now. We also have some testbed sites set up so that QA can run Gherkin tests without having to set up a local environment - the story is the same there, you have to dump the input into a file and then run it which seems unnecessary.

So we have to dump all our tests out from Testrail which isn't that complicated but it would be nice to not have to put everything in a file to run. To get even more ridiculous (but out of scope of this issue) we run all the tests in parallel (with gnu parallel) which requires extracting them even further, then we have feature files looking like this...

```
S719-C12251-restws-automated-scripts-571fc806e9608.feature
S153-C8483-reports-571fc806e6931.feature
S136-C10428-e-commerce-571fc806e288a.feature
S142-C25290-course-objects---requirements-571fc806e383f.feature
S135-C10207-course-enrollments-571fc806e1f10.feature
S142-C12629-course-objects---requirements-571fc806e371e.feature
S142-C25298-course-objects---requirements-571fc806e3b21.feature
```

It "works" but I think we could have something more elegant if I didn't have to make hundreds of files.

**Comment by stof** [comment]
@djdevin you can write a Behat extension which would register a loader reading features from elsewhere than the local filesystem. This is what the (now unmaintained) [Github extension](https://github.com/Behat/GithubExtension) was implementing for Behat 2.4. You could have an extension reading features directly from Testrail

**Comment by djdevin** [comment]
interesting, thanks for the tip!
